testing/scenario_app/ios/Scenarios/ScenariosTests/AppLifecycleTests.m

```diff
@@ -0,0 +1,144 @@
+                                                           @"through AppLifecycleState.inactive"]];
+  [lifecycleExpectations
+      addObject:[[XCTestExpectation alloc]
+                    initWithDescription:
+                        @"A loading FlutterViewController goes through AppLifecycleState.resumed"]];
+
+  FlutterViewController* flutterVC = [rootVC showFlutter];
+  [engine.lifecycleChannel setMessageHandler:^(id message, FlutterReply callback) {
+    if (lifecycleExpectations.count == 0) {
+      XCTFail(@"Unexpected lifecycle transition: %@", message);
+    }
+    [lifecycleEvents addObject:message];
+    [[lifecycleExpectations objectAtIndex:0] fulfill];
+    [lifecycleExpectations removeObjectAtIndex:0];
+  }];
+
+  [self waitForExpectations:lifecycleExpectations timeout:5];
+
+  // Expected sequence from showing the FlutterViewController is inactive and resumed.
+  BOOL lifecyclesAsExpected = [lifecycleEvents
+      isEqualToArray:@[ @"AppLifecycleState.inactive", @"AppLifecycleState.resumed" ]];
+  XCTAssert(lifecyclesAsExpected,
+            @"AppLifecycleState transitions while presenting not as expected");
+
+  // Now dismiss the FlutterViewController again and expect another inactive and paused.
+  [lifecycleExpectations
+      addObject:[[XCTestExpectation alloc]
+                    initWithDescription:@"A dismissed FlutterViewController goes through "
+                                        @"AppLifecycleState.inactive"]];
+  [lifecycleExpectations
+      addObject:[[XCTestExpectation alloc]
+                    initWithDescription:@"A dismissed FlutterViewController goes through "
+                                        @"AppLifecycleState.paused"]];
+  [flutterVC dismissViewControllerAnimated:NO completion:nil];
+  [self waitForExpectations:lifecycleExpectations timeout:5];
+  lifecyclesAsExpected = [lifecycleEvents isEqualToArray:@[
+    @"AppLifecycleState.inactive", @"AppLifecycleState.resumed", @"AppLifecycleState.inactive",
+    @"AppLifecycleState.paused"
+  ]];
+  XCTAssert(lifecyclesAsExpected,
+            @"AppLifecycleState transitions while dismissing not as expected");
+
+  // Now put the app in the background (while the engine is still running) and bring it back to
+  // the foreground. Granted, we're not winning any awards for hyper-realism but at least we're
+  // checking that we aren't observing the UIApplication notifications and double registering
+  // for AppLifecycleState events.
+
+  // However the production is temporarily wrong. https://github.com/flutter/flutter/issues/37226.
+  // It will be fixed in a next PR that removes the wrong asserts.
+  [lifecycleExpectations
+      addObject:
+          [[XCTestExpectation alloc]
+              initWithDescription:@"Current implementation sends another AppLifecycleState event"]];
+  [[NSNotificationCenter defaultCenter]
+      postNotificationName:UIApplicationWillResignActiveNotification
+                    object:nil];
```

Comment:
This test is a bit confusing.  Can you split it into a test per notification with the expectation order enforced instead of the array checks of the messages?

Reply:
It would still make sense to assert the expectations between notifications?  Can you break it down into separate tests at all?

Reply:
I still think this is hard to read and it doesn't actually verify that the expectation was hit because of the notification, because it doesn't wait for the expectations between notifications.  If this test ever fails because something is out of order it's going to be tricky to know which notification went awry and difficult to debug.

Reply:
NSNotificationCenter posts are synchronous, and I'm removing these expectations right after in the next PR.

Reply:
Well you would set up the expectation, post, then wait.  We can talk about it in the next PR.